### ISSUE TITLE
microk8s + ufw behaving weirdly on arm64 vs amd64

### ISSUE THREAD

**Issue by sgasse** [ubuntu/microk8s#2553]
Today I've struggled quite a lot with setting up microk8s on an `arm64` device. My Kubernetes knowledge is rather limited, that is why I was not able to debug it in a good way so far. But maybe someone can nudge me in the right direction?

- When I install microk8s via snap (`sudo snap install microk8s --classic --channel=1.21/stable`) while I have `ufw` enabled, the network interfaces (`vxlan.calico`, `caliXXXXX...`) are not created (checked with `ip a`). If installing or starting (`microk8s start`) with `ufw` disabled, the network interfaces are created.
- When I install microk8s in the same way on a fresh amd64 VM, the network interfaces are created even if `ufw` is enabled.
- Adding the exception rules (in/out on the network interfaces with wildcard, default route policy) does not change this fact on arm64 while the rules can be added after the installation on amd64.

The device is a Nvidia Jetson AGX and it runs a Nvidia-customized Ubuntu 18.04. Since it is configured within a semi-production environment, I cannot post the complete report of `microk8s inspect` without reviewing it. But I could post a part of it if somebody could tell me which part is interesting to pin down this issue.

Any help would be greatly appreciated! :slightly_smiling_face:

**Comment by balchua** [comment]
There was a PR on 1.22 which added the vxlan calico to ufw.
https://github.com/ubuntu/microk8s/pull/2426/files
But I don't think it was added on 1.21.  although i am not sure if this is the issue.
If you don't see the interface it maybe that calico pods didn't startup well.

You can try to see the calico pod logs maybe it can give you some hint.

**Comment by ktsakalozos** [comment]
@sgasse it would help if we could see a `microk8s.inspect` tarball. Do you see the calico pods coming up? Any errors in their logs? Any errors in `dmesg`? How does you `microk8s.kubectl get all -A` look like?

**Comment by sgasse** [comment]
Thank you for the hint with the calico pods @balchua and @ktsakalozos ! :+1: 
I think I solved it now and it (fortunately or unfortunately?) had nothing to do with microk8s itself. I guess there is nothing for microk8s to learn from it, but maybe for other people running into a comparable issue, so I'll describe it below.

**TL;DR**: Missing kernel modules causing issues in `ufw`, solving it by deactivating IPv6 in `/etc/default/ufw`.

When reactivating `ufw` on the `arm64` device again, I noticed an error:
```
$ sudo ufw enable
Command may disrupt existing ssh connections. Proceed with operation (y|n)? y
ERROR: problem running ufw-init
ip6tables-restore: line 142 failed

Problem running '/etc/ufw/before6.rules'
```
I checked the `before6.rules` and even purged+reinstalled `ufw`, since I did not remember tampering with the rules. After some digging around, it seems that some kernel modules are missing on the `arm64` installation, preventing `ufw` from working correctly.

One suggestion was to disable IPv6 if it is not needed. So after switching to
```
IPV6=no
```
in `/etc/default/ufw`, `ufw` can be restarted without error. After this change (and reboot to flush the network interfaces), a clean reinstall of microk8s worked without issues as on `amd64` :tada:

**Comment by sgasse** [comment]
Now I did look at the `calico` pod logs and there is one difference between the `amd64` system and the `arm64` system. The non-controller pod fails on `arm64`, but I am not sure if this is a problem?

On `amd64`:
```
2021-09-04 02:20:38.339 [INFO][10] startup.go 290: Early log level set to info
2021-09-04 02:20:38.339 [INFO][10] startup.go 306: Using NODENAME environment for node name
2021-09-04 02:20:38.340 [INFO][10] startup.go 318: Determined node name: brizo
2021-09-04 02:20:38.341 [INFO][10] startup.go 350: Checking datastore connection
2021-09-04 02:20:38.364 [INFO][10] startup.go 374: Datastore connection verified
2021-09-04 02:20:38.364 [INFO][10] startup.go 102: Datastore is ready
2021-09-04 02:20:38.374 [INFO][10] reachaddr.go 46: Auto-detected address by connecting to remote Destination="<redacted>" IP=<redacted>
2021-09-04 02:20:38.374 [INFO][10] reachaddr.go 55: Checking interface CIDRs Name="vxlan.calico"
2021-09-04 02:20:38.374 [INFO][10] reachaddr.go 57: Checking CIDR CIDR="10.1.184.64/32"
2021-09-04 02:20:38.374 [INFO][10] reachaddr.go 55: Checking interface CIDRs Name="<redacted>"
2021-09-04 02:20:38.374 [INFO][10] reachaddr.go 57: Checking CIDR CIDR="<redacted>"
2021-09-04 02:20:38.374 [INFO][10] reachaddr.go 59: Found matching interface CIDR CIDR="<redacted>"
2021-09-04 02:20:38.374 [INFO][10] startup.go 678: Using autodetected IPv4 address <redacted>, detected by connecting to <redacted>
2021-09-04 02:20:38.374 [INFO][10] startup.go 715: No AS number configured on node resource, using global value
2021-09-04 02:20:38.375 [INFO][10] startup.go 171: Setting NetworkUnavailable to False
2021-09-04 02:20:38.405 [INFO][10] startup.go 764: found v6= in the kubeadm config map
2021-09-04 02:20:38.410 [INFO][10] startup.go 598: FELIX_IPV6SUPPORT is false through environment variable
2021-09-04 02:20:38.420 [INFO][10] startup.go 215: Using node name: <redacted>
2021-09-04 02:20:38.478 [INFO][19] allocateip.go 144: Current address is still valid, do nothing currentAddr="10.1.184.64" type="vxlanTunnelAddress"
CALICO_NETWORKING_BACKEND is vxlan - no need to run a BGP daemon
Calico node started successfully
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x118 pc=0x164c5ae]

goroutine 113 [running]:
github.com/projectcalico/felix/calc.(*VXLANResolver).OnResourceUpdate(0xc00028be00, 0x2123160, 0xc000122900, 0x1d33920, 0x0, 0xc00093cbf0, 0x6, 0x0, 0x0, 0x2, ...)
        /go/pkg/mod/github.com/projectcalico/felix@v0.0.0-20200311115901-aa0a22d97c2d/calc/vxlan_resolver.go:185 +0x37e
github.com/projectcalico/felix/dispatcher.updateHandlers.DispatchToAll(0xc0000ce7a0, 0x1, 0x1, 0x2123160, 0xc000122900, 0x1d33920, 0x0, 0xc00093cbf0, 0x6, 0x0, ...)
        /go/pkg/mod/github.com/projectcalico/felix@v0.0.0-20200311115901-aa0a22d97c2d/dispatcher/dispatcher.go:45 +0x8b
github.com/projectcalico/felix/dispatcher.(*Dispatcher).OnUpdate(0xc0004f7540, 0x2123160, 0xc000122900, 0x1d33920, 0x0, 0xc00093cbf0, 0x6, 0x0, 0x0, 0x2, ...)
        /go/pkg/mod/github.com/projectcalico/felix@v0.0.0-20200311115901-aa0a22d97c2d/dispatcher/dispatcher.go:102 +0x3a2
github.com/projectcalico/felix/dispatcher.(*Dispatcher).OnUpdates(0xc0004f7540, 0xc0009464f0, 0x1, 0x1)
        /go/pkg/mod/github.com/projectcalico/felix@v0.0.0-20200311115901-aa0a22d97c2d/dispatcher/dispatcher.go:80 +0x136
github.com/projectcalico/felix/calc.(*AsyncCalcGraph).loop(0xc00028bec0)
        /go/pkg/mod/github.com/projectcalico/felix@v0.0.0-20200311115901-aa0a22d97c2d/calc/async_calc_graph.go:150 +0x43d
created by github.com/projectcalico/felix/calc.(*AsyncCalcGraph).Start
        /go/pkg/mod/github.com/projectcalico/felix@v0.0.0-20200311115901-aa0a22d97c2d/calc/async_calc_graph.go:237 +0xb3
```
Despite the error, the pod is reported as ready:
```
$ kubectl get pods -A
NAMESPACE     NAME                                      READY   STATUS    RESTARTS   AGE
kube-system   calico-node-xh2xr                         1/1     Running   3          2d17h
kube-system   calico-kube-controllers-f7868dd95-fzzsl   1/1     Running   1          2d17h
```

However on the `arm64` system, the pod is not ready:
```
$ kubectl get pods -A
NAMESPACE     NAME                                      READY   STATUS    RESTARTS   AGE
kube-system   calico-node-gj69w                         0/1     Running   0          29m
kube-system   calico-kube-controllers-f7868dd95-9mjcq   1/1     Running   0          29m
```

```
...
2021-09-06 07:35:47.405 [INFO][33] ipam.go 804: Attempting to assign 1 addresses from block block=10.1.60
.128/26 handle="vxlan-tunnel-addr-amsterdam" host="amsterdam"
2021-09-06 07:35:47.410 [INFO][33] ipam.go 1265: Creating new handle: vxlan-tunnel-addr-amsterdam
2021-09-06 07:35:47.433 [INFO][33] ipam.go 827: Writing block in order to claim IPs block=10.1.60.128/26 
handle="vxlan-tunnel-addr-amsterdam" host="amsterdam"
2021-09-06 07:35:47.458 [INFO][33] ipam.go 840: Successfully claimed IPs: [10.1.60.128/26] block=10.1.60.
128/26 handle="vxlan-tunnel-addr-amsterdam" host="amsterdam"
2021-09-06 07:35:47.458 [INFO][33] ipam.go 569: Auto-assigned 1 out of 1 IPv4s: [10.1.60.128/26] handle="
vxlan-tunnel-addr-amsterdam" host="amsterdam"
2021-09-06 07:35:47.500 [INFO][33] allocateip.go 292: Assigned tunnel address to node IP="10.1.60.128" ty
pe="vxlanTunnelAddress"
CALICO_NETWORKING_BACKEND is vxlan - no need to run a BGP daemon
Calico node started successfully
2021-09-06 07:35:48.795 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1
 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:48.802 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1
 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:48.809 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1
 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:48.819 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:48.833 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:48.857 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:48.896 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:48.967 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:49.104 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:49.370 [ERROR][55] ipsets.go 561: Bad return code from 'ipset list'. error=exit status 1 family="inet" stderr="ipset v6.38: Kernel error received: Invalid argument\n"
2021-09-06 07:35:49.890 [ERROR][55] ipsets.go 916: Failed to read IP sets error=exit status 1 family="inet"
2021-09-06 07:35:49.890 [PANIC][55] ipsets.go 342: Failed to update IP sets after multiple retries. family="inet"
...
```

**Comment by ktsakalozos** [comment]
Maybe another module is missing? `eg modprobe xt_set`

**Comment by sgasse** [comment]
Looks like you are right about missing kernel modules @ktsakalozos , thank you! It turns out that the Jetson devices would require compiling a custom kernel for that. The solution was disabling `ha-cluster` and [using flannel instead of calico](https://github.com/ubuntu/microk8s/issues/2110#issuecomment-803674972).

Though given that the GPU operator does not work with integrated GPUs (like the Tegra on Jetson devices) and `microk8s` switched to `containerd` since `1.14`, closing the possibility to use `docker` with the `nvidia` runtime, we unfortunately will have to go with Kubernetes from scratch, as part of the [EGX Stack](https://github.com/NVIDIA/egx-platform/blob/master/install-guides/Jetson_Xavier_NX_v2.0.md)

Thank you anyway for the help, I understand the ins and outs of it a lot better now! :slightly_smiling_face:
